components/resource_control/src/service.rs

```diff
@@ -0,0 +1,272 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{sync::Arc, time::Duration};
+
+use futures::StreamExt;
+use kvproto::{pdpb::EventType, resource_manager::ResourceGroup};
+use pd_client::{Error as PdError, PdClient, Result, RpcClient, RESOURCE_CONTROL_CONFIG_PATH};
+use tikv_util::{box_err, error};
+
+use crate::ResourceGroupManager;
+
+#[derive(Clone)]
+pub struct ResourceManagerService {
+    manager: Arc<ResourceGroupManager>,
+    pd_client: Arc<RpcClient>,
+    // record watch revision
+    revision: i64,
+}
+
+impl ResourceManagerService {
+    /// Constructs a new `Service` with `ResourceGroupManager` and a `RpcClient`
+    pub fn new(
+        manager: Arc<ResourceGroupManager>,
+        pd_client: Arc<RpcClient>,
+    ) -> ResourceManagerService {
+        ResourceManagerService {
+            pd_client,
+            manager,
+            revision: 0,
+        }
+    }
+}
+
+impl ResourceManagerService {
+    pub async fn watch_resource_groups(&mut self) {
+        // Firstly, load all resource groups as of now.
+        match self.list_resource_groups().await {
+            Ok((groups, revision)) => {
+                self.revision = revision;
+                for group in groups {
+                    self.manager.add_resource_group(group);
+                }
+            }
+            Err(err) => error!("failed to list resource groups, err: {:?}", err),
```

Comment:
Should also retry `list_resource_groups` here

Reply:
Okay, i will put all in loop